### PR TITLE
correctness build optimizations

### PR DESCRIPTION
**Pull request by jmarolf**
depends on https://github.com/dotnet/roslyn/pull/628

this PR does three things:

1. only run `Correctness_Determinism`, `Correctness_Build`, and `Correctness_Rebuild` on check-in rather than PRs
2. rename `Correctness_Build` to `Correctness_Bootstrap_Build`
3. Create 2 new jobs (`Correctness_Code_Analysis` and `Correctness_Build_Artifacts`) that validate everything the `Correctness_Bootstrap_Build` job does but without using the boot strap compiler.

## Motivation

The bootstrap compiler is _very_ slow and running it on all of Roslyn takes well over an hour. There are bugs that doing this will catch but I am proposing in this PR that we do not block merging on them. With this change we can still validate that we are not breaking bootstrapping, determinism, re-build, or hitting asserts in the analyzer driver. We will just get that notification on checkin.

### CODE REVIEW

File: eng/pipelines/build-windows-job.yml [hunk 1]
```diff
@@ -61,7 +61,7 @@ jobs:
       displayName: Publish Test Payload
       inputs:
         targetPath: '$(Build.SourcesDirectory)\artifacts\testPayload'
-        artifactName: $artifactName
+        artifactName: ${{ parameters.testArtifactName }}
```
Comment (JoeRobich) [inline eng/pipelines/build-windows-job.yml:64]: good catch!

**Review by 333fred — CHANGES_REQUESTED** [review]
I have serious concerns about this, in particular for any code that touches the compiler leg. Yes, the bootstrap build is slow. The bugs it catches are the types of bad bugs that will introduce significant issues when we attempt to insert roslyn. I would be ok with not running the bootstrap on IDE-only prs, but if you touch the compiler, I believe it needs to stay.

**Comment by jmarolf** [comment]
> I have serious concerns about this, in particular for any code that touches the compiler leg. Yes, the bootstrap build is slow. The bugs it catches are the types of bad bugs that will introduce significant issues when we attempt to insert roslyn. 

An alternative tactic could be to come up with a smaller project to build. Building 5+ million lines with a debug build feels like a stress test instead of actually validating functionality. Yes, it has caught important bugs but couldn't we catch those same bugs with a smaller project size?

> I would be ok with not running the bootstrap on IDE-only prs, but if you touch the compiler, I believe it needs to stay.

I'll investigate only running these if files in `src/Compilers` have changed.

**Comment by jaredpar** [comment]
> > I have serious concerns about this, in particular for any code that touches the compiler leg. 
> I'll investigate only running these if files in src/Compilers have changed.

Overall I'm okay with us moving some of checks to check-in vs. PR so long as there as a concrete person who is on the hook for monitoring those runs. Basically if the tests fail in CI we know a bug / email will be raised within say 24 hours. That is a reasonable trade off for this type of check. One where it finds real bugs, has an extremely low false positive rate but also only finds ~1 bug a month. 

At the same time some changes are more suspect than others. A change to `src/Compilers` is a lot more suspicious than anything in the IDE. Having it run always for `src/Compilers` but CI only for everything else is a good compromise for our first steps here. 

@jmarolf the dotnet/runtime team has infra in place for having legs run based on which files were changed. They had to do similar subsetting for changes to CLR vs. libraries. Likely we could re-use much of what they put into place. 

Note: really I wish there was a way we could flag PR messages with say `/gh run compilers` so we could just dynamically add legs to PRs vs. looking for file changes.

**Comment by 333fred** [comment]
> An alternative tactic could be to come up with a smaller project to build.

I would probably be fine with just bootstrapping Compilers.sln.

**Comment by JoeRobich** [comment]
To only run these when compiler changes are made, we may need to break them into a new pipeline as path triggers are at the pipeline level not the job level.

https://github.com/dotnet/roslyn/blob/d72982adf0bdac7f44a95f94912cfa7ddacde73f/azure-pipelines.yml#L10-L29

**Comment by RikkiGibson** [comment]
we could also consider inserting a prereq job for the correctness jobs which runs on the hosted queues (i.e. has a fast queue time) and simply searches the set of changed paths for anything under src/Compilers.

The Correctness jobs don't and probably can't depend on the test payload, so there's probably not much risk in breaking them out into their own pipeline. It is more stuff to monitor and keep track of, though.

**Comment by jaredpar** [comment]
> I would probably be fine with just bootstrapping Compilers.sln.

I would be fine for that on PR but for CI would want to do the entire repo. Historically their have been a few pieces of IDE code that triggered Compiler bugs.

**Comment by jaredpar** [comment]
> To only run these when compiler changes are made, we may need to break them into a new pipeline as path triggers are at the pipeline level not the job level.

The way the runtime team approached this is they had a task at the start which just looked effectively looked at the changes in the PR. Then the expported variables like `CHANGES_CLR=true`, `CHANGES_LIBRARIES=false`. Later they conditioned the tasks that did library only tests on those variables.

**Comment by jmarolf** [comment]
@jaredpar can I get another review pass now that I've implemented this running the bootstrap build if there are compiler changes?

File: azure-pipelines.yml [hunk 2]
```diff
@@ -180,14 +184,71 @@ jobs:
...
+      displayName: Validate Build Artifacts
+
+    - task: PowerShell@2
+      displayName: Generate Syntax Files
+      inputs:
+        filePath: eng/generate-compiler-code.ps1
+        arguments: -test -configuration Release
+        
+    - script: $(Build.SourcesDirectory)\.dotnet\dotnet.exe tool run dotnet-format whitespace  $(Build.SourcesDirectory)\src --folder --include-generated --include $(Build.SourcesDirectory)\src\Compilers\CSharp\Portable\Generated\ $(Build.SourcesDirectory)\src\Compilers\VisualBasic\Portable\Generated\ $(Build.SourcesDirectory)\src\ExpressionEvaluator\VisualBasic\Source\ResultProvider\Generated\ --verify-no-changes
+      displayName: Validate Generated Syntax Files
```
Comment (333fred) [inline azure-pipelines.yml:235]: As long as we're trying to optimize this leg, should we avoid this step for IDE PRs?

File: azure-pipelines.yml [hunk 3]
```diff
@@ -196,7 +257,9 @@ jobs:
         jobName: Correctness_Determinism
         configuration: Debug
 
-- job: Correctness_Build
+- job: Correctness_Bootstrap_Build
+  dependsOn: Determine_Changes
+  condition: eq(dependencies.Determine_Changes.outputs['SetPathVars_compilers.containsChange'], 'true')
```
Comment (333fred) [inline azure-pipelines.yml:262]: Is this filter only for PR runs?

Comment (jmarolf) [reply azure-pipelines.yml:262]: nope let me fix that

File: azure-pipelines.yml (hunk 2 above)
Comment (jmarolf) [reply azure-pipelines.yml]: Done

File: azure-pipelines.yml (hunk 3 above)
Comment (jmarolf) [reply azure-pipelines.yml]: done

File: azure-pipelines.yml [hunk 4]
```diff
@@ -184,14 +184,68 @@ jobs:
...
+        filePath: eng/build.ps1
+        arguments: -configuration Release -prepareMachine -ci -restore -binaryLog
+
+    - task: PowerShell@2
+      displayName: Build
+      inputs:
+        filePath: eng/build.ps1
+        arguments: -configuration Release -prepareMachine -ci -build -pack -publish -sign -binaryLog
+
+    - script: $(Build.SourcesDirectory)\artifacts\bin\BuildBoss\Release\net472\BuildBoss.exe  -r "$(Build.SourcesDirectory)/" -c Release -p Roslyn.sln
```
Comment (333fred) [inline azure-pipelines.yml:220]: Shouldn't we run BuildBoss on most PRs? It verifies more than compiler packages.

Comment (jmarolf) [reply azure-pipelines.yml:220]: I should always run. I believe it doesn't have any conditions on it

**Review by 333fred — APPROVED** [review]

**Comment by jmarolf** [comment]
verified these changes in https://github.com/dotnet/roslyn/pull/63368

**Pull request merged**